### PR TITLE
fix(addon): protect modifier shortcuts and configure hold delay

### PR DESCRIPTION
**Pull request by fanzhuyifan**
Fix modifier-only shortcuts firing during combinations such as Ctrl+C. Dictation, command mode, and the palette share tap/hold recognition, with a configurable hold delay (300 ms by default). Single-modifier bindings consume both press and release; multi-modifier chords forward their events, preserving individual modifier use and avoiding unmatched presses. Applications may also respond to those chords. Auto-repeat no longer restarts pending hold activation.

Another key cancels a pending gesture or discards a still-held modifier-only recording. Recording and post-processing share cancellation. Active-recording cancellation for other shortcuts is deferred; mouse clicks are outside scope.

Validation: all 12 CTest groups pass, including 3,796 gesture scenarios and 16 application-routing cases with text-delivery, cancellation, and hold-repeat checks. Formatting and i18n checks pass. Remote CI and live desktop testing remain unverified.

## Summary by CodeRabbit

- **New Features**
  - Added a configurable hold activation delay, defaulting to 300 ms.
  - Improved modifier-key shortcuts for taps, holds, cancellation, and multi-key combinations.
  - Improved recording cancellation, including discarding interrupted recordings.

- **Documentation**
  - Updated English and Chinese guides and configuration references with shortcut behavior and configuration details.
  - Added translated help text for the hold activation delay.

- **Tests**
  - Added automated coverage for shortcuts, configuration, and cancellation.
  - CI now runs the test suite after building.

### Follow-up implementation
- [x] Activate modifier taps on the first required release, rearm repeated chords, and update regression tests and shortcut documentation.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -69,6 +69,11 @@ jobs:
       - name: Build all targets
         run: cmake --build --preset ci-clang-mold
 
+      - name: Run tests
+        run: |
+          command -v dbus-run-session
+          ctest --test-dir build --output-on-failure
```
[the remark this answers is not in the document]
Comment (fanzhuyifan) [reply .github/workflows/ci.yml:75]: I will let the maintainer decide how to run the CI and whether the these added tests are reasonable...

File: src/daemon/main.cpp [hunk 2]
```diff
@@ -586,6 +586,8 @@ int main(int argc, char* argv[]) {
   dbus.SetCancelPostprocessingHandler([&](bool commit_raw_text) {
     return runtime_controller.CancelPostprocessing(commit_raw_text);
   });
+  dbus.SetCancelOperationHandler(
+      [&](bool commit_raw_text) { return runtime_controller.CancelOperation(commit_raw_text); });
```
[the remark this answers is not in the document]
Comment (fanzhuyifan) [reply src/daemon/main.cpp:587]: Documented CancelPostprocessing as deprecated and directed new callers to CancelOperation. Kept the legacy D-Bus endpoint and its post-processing-only behavior for compatibility. The internal implementation remains shared with CancelOperation.

Also happy to just remove that..

**Review by xifan2333 — COMMENTED** [review]
Hi @fanzhuyifan, thanks for the hard work on the modifier gesture engine and the comprehensive test suite!

The overall gesture logic and test coverage look very solid. Having single-modifier shortcuts consume both press and release works well for dedicated bindings like `Shift_R` (leaving `Shift_L` for standard IME toggling without accidental double-triggers).

Before merging, there are just a few engineering cleanups needed:

1. **Clean up legacy `CancelPostprocessing` D-Bus endpoint**:
   Since `fcitx5-vinput` is an internal full-stack component (addon and daemon are built and released together) rather than a public library, we do not need to keep deprecated endpoints for backward compatibility. Since the addon has fully migrated to `CancelOperation`, please remove `CancelPostprocessing` completely from `dbus_interface.h`, `dbus_service`, `main.cpp`, and the tests.

2. **Explicit CI dependency**:
   Please add `dbus-daemon` to `scripts/ci-install-build-deps.sh` under `ubuntu24.04` and `debian12` so that hermetic test environments have `dbus-run-session` explicitly available and CodeRabbit noise is eliminated.

3. **Resolve `cpp-linter` diagnostics**:
   The `cpp-linter` workflow currently reports clang-tidy diagnostics (mostly missing standard includes like `<vector>`, `<utility>`, casting warnings, etc.). Running `mise run check:changed` locally should help identify and resolve them.

Once CI turns green, we'll verify this on a live desktop session and proceed with the merge!

**Comment by xifan2333** [comment]
Hi @fanzhuyifan,

Thank you again for your hard work on this PR! The extensive test cases and thorough exploration of modifier edge cases have been extremely valuable in clarifying all the requirements here.

After reviewing the architecture and exploring Fcitx5's core codebase, our strong preference is to implement this feature by aligning with **upstream Fcitx5's idiomatic event patterns** rather than maintaining a custom standalone gesture state machine:
- [`fcitx5/src/lib/fcitx/instance.cpp#L766-L880`](https://github.com/fcitx/fcitx5/blob/master/src/lib/fcitx/instance.cpp#L766-L880)
- [`fcitx5/src/lib/fcitx/misc_p.h#L266-L273`](https://github.com/fcitx/fcitx5/blob/master/src/lib/fcitx/misc_p.h#L266-L273)

In upstream's pattern:
1. **On modifier press** (L855–L878): It records the pressed key and timestamp, then calls `keyEvent.filter()` (*"don't forward to input method, but make it pass through to client"*). This preserves modifier delivery to apps while naturally preventing combos (like `Ctrl+C` or `Shift+A`) from triggering single-modifier actions.
2. **On modifier release** (L818–L843): It uses Fcitx5's built-in `origKey.isReleaseOfModifier(lastKeyPressed)` with a timeout to detect single-key taps.
3. **Timer management**: A single `EventLoop` time event handles hold delay/activation cleanly, removing the need for parallel timer systems.

**Why we favor this direction:**
- It unifies modifier and non-modifier shortcuts into a single, clean event path in `vinput_keyevent.cpp` (~50–80 lines of core logic instead of maintaining an external recognizer).
- It removes the manual key queues (`PressedKey`, `consumed_`) and duplicate timers, ensuring seamless integration with the rest of Fcitx5.
- The great test scenarios you've already written can directly serve as the test suite for this native implementation!

Would you be interested in refactoring the gesture handling toward this upstream-aligned approach? If you'd like, we can also collaborate or provide sample implementations. 

Let us know what you think!

**Comment by fanzhuyifan** [comment]
> Would you be interested in refactoring the gesture handling toward this upstream-aligned approach? If you'd like, we can also collaborate or provide sample implementations.

Let me check this out... Often modifier-only shortcuts require special treatment.

**Comment by fanzhuyifan** [comment]
Thanks for the suggestion! I looked into the upstream implementation. Reusing Fcitx’s helpers and event-loop facilities makes sense, but adopting its shortcut behavior wholesale would change several intentional behaviors covered by this PR.

| Aspect | Current PR | Adopting upstream shortcut behavior directly |
| --- | --- | --- |
| Standalone modifier delivery | Currently consumes the configured key’s presses and releases; **I’m reconsidering this and would prefer pass-through if we can resolve the Firefox issue below** | Passes them through to the application, which may also respond to the same modifier |
| Modifier chord press order | Recognizes the chord regardless of press order | A single binding is order-sensitive: `Control+Shift_L` matches Shift_L pressed while Ctrl is already down; reverse order requires an additional binding |
| Chord tap activation | **Now activates on the first required modifier release**, with rearming while other modifiers remain held | Can activate on the first matching modifier release |
| Tap timeout | Dictation and command shortcuts in Tap mode have no maximum hold duration | Defaults to 250 ms; configurable, with `-1` disabling the timeout |
| Modifier hold behavior | Supports configurable activation delay, release-to-stop, and interruption-to-discard | The referenced shortcut handler does not provide these recording-specific behaviors; we would still need to implement them |

These differences follow from the upstream [shortcut handler](https://github.com/fcitx/fcitx5/blob/c7ecdb931d8b378ccdcd87382a3ef7ff0bd10def/src/lib/fcitx/instance.cpp#L813-L880), [`Key::isReleaseOfModifier()` and key-matching implementation](https://github.com/fcitx/fcitx5/blob/c7ecdb931d8b378ccdcd87382a3ef7ff0bd10def/src/lib/fcitx-utils/key.cpp#L335-L412), and [timeout configuration](https://github.com/fcitx/fcitx5/blob/c7ecdb931d8b378ccdcd87382a3ef7ff0bd10def/src/lib/fcitx/globalconfig.cpp#L164-L176).

There are practical reasons to preserve several of the current behaviors, though I’m reconsidering modifier event delivery:

- **Modifier event delivery:** On reconsideration, preserving normal modifier delivery to applications seems preferable, and Fcitx’s `filter()` pattern makes sense here. With an earlier pass-through implementation, however, voice input was unusable in Firefox on my setup (**Arch Linux, KDE Plasma, Wayland**). I haven’t isolated the cause, so I don’t think that observation alone establishes that consuming modifier events is necessary. I’d like to investigate it and use pass-through if we can make it work reliably.
- **Order-independent chord recognition:** Users pressing Ctrl+Shift should not have to ensure that Ctrl physically goes down before Shift every time, especially when pressing the keys nearly simultaneously.
- **No maximum duration for voice taps:** In Tap mode, users should not have to release the shortcut quickly for it to register. Older users and people with motor impairments may need more time to press and release keys comfortably. Fcitx already supports unlimited modifier taps through `ModifierOnlyKeyTimeout=-1`; my preference is for voice taps to allow this without requiring users to change the global timeout for other Fcitx shortcuts. Hold/Both modes retain their configurable hold threshold.
- **Explicit modifier hold and cancellation behavior:** The activation delay avoids starting recordings during brief modifier use. Releasing a required modifier ends the hold, and pressing another key during an active hold cancels and discards the recording so ordinary shortcut use does not leave an unintended voice operation running (e.g., pressing Ctrl+C but having a long delay between Ctrl and C should cancel the recording if Ctrl is the shortcut for recording).

I’ve now implemented **first-release activation**, including rearming chords while other modifiers remain held. I forgot KDE [also adopted this](https://github.com/KDE/kglobalacceld/commit/7a6840180e5b710afa59fc297aabddf727173627) after [the initial MR](https://invent.kde.org/plasma/kglobalacceld/-/merge_requests/44). It makes repeated chords easier—for example, keeping Ctrl held while repeatedly tapping Shift—and removes the release-window timing constraint.

Modifier chord recognition and tap activation timing are now broadly aligned with KDE. Event consumption currently differs and is under reconsideration; mouse interaction and recording-specific hold behavior also remain different.

I’m happy to reuse upstream helpers where they simplify the code. We still need state for chord recognition, interruption, and repeat handling, but passing modifier events through consistently may let us simplify or remove the consumed-key ownership tracking.

Another possibility is to contribute focused improvements upstream, particularly making a single modifier-only binding recognize either press order.

Happy to hear your thoughts on all this.

**Comment by fanzhuyifan** [comment]
Another point of discussion is whether we want to make modifier-only shortcuts configured for "both" activate on press or release. The pro for activating on press is that this is consistent with other shortcuts and that the user experiences no delay, the con is that even though this could be canceled on other key press, we may still see an annoying popup that says ...recording... every time we press this...

**Comment by xifan2333** [comment]
Hi @fanzhuyifan, thanks again for digging into the modifier-only shortcut behavior and especially for calling out the Firefox/KDE/Wayland pass-through concern.

We ended up implementing this separately in #207 with a smaller upstream-aligned path rather than carrying a standalone gesture recognizer:

- modifier press uses Fcitx's `filter()` pass-through pattern, so combinations such as `Alt+Tab`, `Ctrl+C`, and `Shift+A` are left to the client/application instead of being consumed by vinput;
- modifier release uses Fcitx's `Key::isReleaseOfModifier()` style disambiguation before treating it as a single-modifier tap;
- hold activation uses the Fcitx event loop with the configurable `HoldActivationDelay`;
- recording-specific behavior is still kept at the action layer: short tap starts/toggles voice input, hold starts after the delay and stops on release, and interrupted hold recordings are cancelled/discarded.

One important scope point: this is still handled through Fcitx `InputContextKeyEvent`, not a desktop-global shortcut grab. So in practice vinput only participates where a frontend/focused input context delivers key events to Fcitx (e.g. editable text fields). Outside that IME/input-context path, Firefox/KDE/application shortcuts remain application-side. During local smoke testing on Arch + KDE Plasma + Wayland + Firefox, `Alt+E` and similar Firefox/app combinations reached Firefox and did not trigger vinput; unbound `Alt+...` behavior can still be Firefox's own menu handling, but vinput should not pop recording/palette UI for those combinations.

We also ran the release workflow dry-run successfully for the current main artifacts. We'll keep validating the Firefox case with the packaged Arch build before tagging.

*AI-assisted — Tool: pi; model: openai-codex/gpt-5.5; version: 0.85.1.*

**Comment by fanzhuyifan** [comment]
Thanks! In the new PR, I cannot reproduce the problem I had with firefox. While having unbalanced press/release events may cause stuck modifiers, maybe we can worry about that once people report it. I do have some comments on https://github.com/xifan2333/fcitx5-vinput/pull/207, and I will move the discussion there.
